### PR TITLE
#12697 - Assign case(s) to a User and allow them to see the data of o…

### LINKED ISSUE #12697: Assign case(s) to a User and allow them to see the data of only the assigned case(s) in the system [8]
### Feature Description

As a system administrator in SORMAS I want to assign a user a role/right so that the user can only view and manage/edit the case(s) and associated data of the case(s) (for example contacts, samples, person, etc.) that are assigned to them.  This should be possible irrespective of jurisdiction, facility, or disease type of both the case and assigned user.

### Added Value/Benefit

This approach is particularly useful during outbreak investigations where temporary, external specialists are brought in to assist a team/country.  They should not see data from other cases/contacts/event participants in the system per Data Protection Officer (DPO) requirements.

For example, a temporary employee can only view (in the Case Directory) and edit the 5 cases which are assigned (through Tasks) to them.

### Acceptance Criteria

- [ ] A new boolean `restrictAccessToAssignedEntities` field has been added to the UserRole entity
  - Caption: "Restrict access to assigned entities"
  - Description: "User can only see entities (e.g. cases, contacts) they have been assigned to as the responsible user. Other entities will not be visible to them, even if they are in their jurisdiction."
- [ ] The new field has been added as a checkbox below the "Port health user" checkbox in the user role edit form
- [ ] A checkbox filter has been added to the user roles directory that, if checked, only shows user roles for which this field is enabled ("Only show user roles with access restricted to assigned entities")
- [ ] A similar filter has been added to the user management directory ("Only show users with access restricted to assigned entities")
- [ ] All user filters and jurisdiction logic has been extended to take into account the new field
  - Entities that a user can be assigned to as responsible user include cases (surveillanceOfficer or caseOfficer), events, environments and contacts (contactOfficer); please check if something's missing from this list
  - The expected outcome is that users with this access restriction ONLY see entities they have been assigned to as the responsible user, INCLUDING all associated entities; e.g. a user who has been assigned as the surveillance officer to a case would see all contacts, tasks, samples, etc. of this case as long as the normal jurisdiction filter allows it (or they have been assigned to the child entity, e.g. the contact, as well)
  - The user still sees entities they have created (i.e. they are the reporting user of)
- [ ] It's verified that the synchronization of data to the mobile device adheres to this new logic

### Implementation Details

_No response_

### Mockups

_No response_

### Additional Information

Possibly related to:
https://github.com/SORMAS-Foundation/SORMAS-Project/issues/5579
https://github.com/SORMAS-Foundation/SORMAS-Project/issues/11330
https://github.com/SORMAS-Foundation/SORMAS-Project/issues/898
https://github.com/SORMAS-Foundation/SORMAS-Security/issues/7

### PR DESCRIPTION
**Pull request by sergiupacurariu**
…nly the assigned case(s) in the system

Fixes #12697

### CODE REVIEW

File: sormas-backend/src/main/java/de/symeda/sormas/backend/user/UserFacadeEjb.java [hunk 1]
```diff
@@ -737,7 +742,8 @@ public List<UserDto> getIndexList(UserCriteria userCriteria, Integer first, Inte
 
 		cq.select(user);
 
-		return QueryHelper.getResultList(em, cq, first, max, UserFacadeEjb::toDto);
+		final List<UserDto> resultList = QueryHelper.getResultList(em, cq, first, max, UserFacadeEjb::toDto);
+		return resultList.stream().distinct().collect(Collectors.toList());
```
Comment (leventegal-she) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/user/UserFacadeEjb.java:746]: please add a comment why`.stream().distinct()` needed, and don't forget to create an issue for improvung this

File: sormas-backend/src/main/resources/sql/sormas_schema.sql [hunk 2]
```diff
@@ -12828,4 +12828,11 @@ ALTER TABLE manualmessagelog_history
 INSERT INTO schema_version (version_number, comment)
 VALUES (535, 'Display a history of sent external emails #12465');
 
+-- 2023-12-05 Assign case(s) to a User and allow them to see the data of only the assigned case(s) in the system #12697
+ALTER TABLE userroles ADD COLUMN restrictAccessToAssignedEntities boolean NOT NULL DEFAULT false;
+ALTER TABLE userroles_history ADD COLUMN restrictAccessToAssignedEntities boolean;
+
+INSERT INTO schema_version (version_number, comment) VALUES (535, 'Add tested pathogen details #12663');
```
Comment (leventegal-she) [inline sormas-backend/src/main/resources/sql/sormas_schema.sql:12835]: the comment is wrong

**Review by leventegal-she — CHANGES_REQUESTED** [review]

File: sormas-backend/src/main/resources/sql/sormas_schema.sql (hunk 2 above)
Comment (sergiupacurariu) [reply sormas-backend/src/main/resources/sql/sormas_schema.sql:12835]: changed

File: sormas-backend/src/main/java/de/symeda/sormas/backend/environment/EnvironmentService.java [hunk 3]
```diff
@@ -90,36 +90,43 @@ public Predicate createUserFilter(EnvironmentQueryContext queryContext) {
...
-			case COMMUNITY:
-				if (currentUser.getCommunity() != null) {
-					filter = CriteriaBuilderHelper
-						.or(cb, filter, cb.equal(environmentJoins.getLocation().get(Location.COMMUNITY), currentUser.getCommunity()));
+		if (currentUserHasRestrictedAccessToAssignedEntities()) {
+			filter =
+				CriteriaBuilderHelper.and(cb, filter, cb.equal(environmentJoin.get(Environment.RESPONSIBLE_USER).get(User.ID), currentUser.getId()));
+		}
+
+		if (!currentUserHasRestrictedAccessToAssignedEntities()) {
```
[the remark this answers is not in the document]
Comment (sergiupacurariu) [reply sormas-backend/src/main/java/de/symeda/sormas/backend/environment/EnvironmentService.java:98]: changed

File: sormas-backend/src/main/java/de/symeda/sormas/backend/environment/environmentsample/EnvironmentSampleJurisdictionValidator.java [hunk 4]
```diff
@@ -45,13 +46,30 @@ public static EnvironmentSampleJurisdictionValidator of(EnvironmentSampleQueryCo
 		return new EnvironmentSampleJurisdictionValidator(qc.getCriteriaBuilder(), user, qc.getJoins(), null);
 	}
 
-	@Override
-	public Predicate isRootInJurisdictionOrOwned() {
-		final Predicate reportedByCurrentUser = cb.and(
+	public Predicate isRootInJurisdictionOrOwned(boolean currentHasUserRestrictedAccess) {
+		if (currentHasUserRestrictedAccess) {
+			final Predicate reportedByCurrentUser = getReportedByCurrentUserPredicate();
+			final Predicate restrictedAccess = currentHasUserRestrictedAccess
```
[the remark this answers is not in the document]
Comment (sergiupacurariu) [reply sormas-backend/src/main/java/de/symeda/sormas/backend/environment/environmentsample/EnvironmentSampleJurisdictionValidator.java:52]: should refactor again after change in logic, done :)

File: sormas-backend/src/main/java/de/symeda/sormas/backend/event/EventService.java [hunk 5]
```diff
@@ -466,62 +466,69 @@ public Predicate createUserFilter(final EventQueryContext queryContext, final Ev
...
-				break;
-			case HEALTH_FACILITY:
-				if (currentUser.getHealthFacility() != null && currentUser.getHealthFacility().getDistrict() != null) {
-					filter = CriteriaBuilderHelper
-						.or(cb, filter, cb.equal(eventJoins.getLocation().get(Location.DISTRICT), currentUser.getHealthFacility().getDistrict()));
+		if (currentUserHasRestrictedAccessToAssignedEntities()) {
+			filter = CriteriaBuilderHelper.and(cb, filter, cb.equal(eventJoin.get(Event.RESPONSIBLE_USER).get(User.ID), currentUser.getId()));
+		}
+
+		if (!currentUserHasRestrictedAccessToAssignedEntities()) {
```
[the remark this answers is not in the document]
Comment (sergiupacurariu) [reply sormas-backend/src/main/java/de/symeda/sormas/backend/event/EventService.java:473]: changed

File: sormas-backend/src/main/java/de/symeda/sormas/backend/user/CurrentUserService.java [hunk 6]
```diff
@@ -98,6 +99,20 @@ public boolean hasAnyUserRight(Set<UserRight> userRights) {
 		return getCurrentUser().hasAnyUserRight(userRights);
 	}
 
+	public boolean hasRestrictedAccessToAssignedEntities() {
+		if (getCurrentUser() != null && !getCurrentUser().getUserRoles().isEmpty()) {
+			if (!getCurrentUser().getUserRoles()
```
[the remark this answers is not in the document]
Comment (sergiupacurariu) [reply sormas-backend/src/main/java/de/symeda/sormas/backend/user/CurrentUserService.java:104]: consistent change here, turned out very nice and short :)

File: sormas-backend/src/main/java/de/symeda/sormas/backend/user/UserFacadeEjb.java (hunk 1 above)
Comment (sergiupacurariu) [reply sormas-backend/src/main/java/de/symeda/sormas/backend/user/UserFacadeEjb.java:746]: added a comment

**Review by leventegal-she — APPROVED** [review]

**Pull request merged**